File: 05.js
// read input
const fs = require("fs");
const input = fs.readFileSync('05.txt', 'utf8').split("\n");

// part 1
const nice1 = a => a.replace(/[^aeiou]+/g, "").length >= 3;
const nice2 = a => a.match(/([a-zA-Z0-9])\1{1}/) !== null;
const nice3 = a => a.match(/ab|cd|pq|xy/) === null;

console.log(
  "Part 1:",
  input
    .filter(v => nice1(v) && nice2(v) && nice3(v))
    .length
);

// part 2
const nice4 = a => a.match(/([a-zA-Z0-9]{2}).*\1{1}/) !== null;
const nice5 = a => a.match(/([a-zA-Z0-9]).{1}\1/) !== null;

console.log(
  "Part 2:",
  input
    .filter(v => nice4(v) && nice5(v))
    .length
);

File: 05.R
# read & parse input
input <- readLines("05.txt", warn = FALSE)

# part 1
nice1 <- function(x) nchar(gsub("[^aeiou]", "", x)) >= 3
nice2 <- function(x) grepl("([a-zA-Z0-9])\\1{1}", x)
nice3 <- function(x) !grepl("ab|cd|pq|xy", x)

# part 2
nice4 <- function(x) grepl("([a-zA-Z0-9]{2}).*\\1{1}", x, perl = TRUE)
nice5 <- function(x) grepl("([a-zA-Z0-9]).{1}\\1", x)

print(paste0(
  "Part 1: ",
  sum(
    vapply(input, function(x) nice1(x) && nice2(x) && nice3(x), logical(1))
  )
))
print(paste0(
  "Part 2: ",
  sum(
    vapply(input, function(x) nice4(x) && nice5(x), logical(1))
  )
))
